OpenRA.Mods.Common/Traits/Player/ParallelProductionQueue.cs

```diff
@@ -0,0 +1,73 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ParallelProductionQueueInfo : ProductionQueueInfo
+	{
+		public override object Create(ActorInitializer init) { return new ParallelProductionQueue(init, init.Self.Owner.PlayerActor, this); }
+	}
+
+	public class ParallelProductionQueue : ProductionQueue
+	{
+		public ParallelProductionQueue(ActorInitializer init, Actor playerActor, ParallelProductionQueueInfo info)
+			: base(init, playerActor, info) { }
+
+		protected override void TickInner(Actor self, bool allProductionPaused)
+		{
+			if (!allProductionPaused)
+			{
+				for (var i = 0; i < Queue.Count;)
+				{
+					var first = Queue[0];
+					var before = first.RemainingTime;
+
+					if (!first.Paused)
+						first.Tick(playerResources);
+
+					foreach (var item in Queue.FindAll(a => a.Item == first.Item))
+					{
+						i++;
+						Queue.Remove(item);
+						Queue.Add(item);
+					}
+
+					if (first.RemainingTime != before)
+						return;
+				}
+			}
+		}
+
+		public override bool IsProducing(ProductionItem item)
+		{
+			return Queue.Contains(item);
+		}
+
+		public override int RemainingTimeActual(ProductionItem item)
+		{
+			var parallelBuilds = 0;
+			var added = new HashSet<string>();
```

Comment:
RemainingTimeActual is called a *lot*, so you ideally want to avoid allocating objects here.  I won't block review over this as this trait isn't used in the default mods, but you should still try and fix this.